File: summary-report-generator.js
const fs = require('fs');
const path = require('path');

/**
 * Generate a comprehensive summary report of the Canvas-Discourse integration
 */
async function generateSummaryReport() {
  console.log('Generating summary report...');
  
  // Load configuration
  const configPath = path.join(__dirname, 'config.json');
  const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
  
  // Load conflict data if available
  let conflicts = [];
  const conflictsPath = path.join(config.paths.analysis, 'conflicts', 'port_conflicts.md');
  if (fs.existsSync(conflictsPath)) {
    conflicts = extractConflictsFromMarkdown(conflictsPath);
  }
  
  // Create report content
  const content = `# Canvas-Discourse Integration Summary Report

*Generated on ${new Date().toLocaleDateString('en-US', { year: 'numeric', month: 'long', day: 'numeric' })}*

## Integration Status

This report provides an executive summary of the Canvas-Discourse LMS integration project.

### Overall Completion

| Component | Canvas | Discourse | Overall |
|-----------|--------|-----------|---------|
| Models | ${config.port.canvas.completion.models}% | ${config.port.discourse.completion.models}% | ${Math.round((config.port.canvas.completion.models + config.port.discourse.completion.models) / 2)}% |
| Controllers | ${config.port.canvas.completion.controllers}% | ${config.port.discourse.completion.controllers}% | ${Math.round((config.port.canvas.completion.controllers + config.port.discourse.completion.controllers) / 2)}% |
| Services | ${config.port.canvas.completion.services}% | ${config.port.discourse.completion.services}% | ${Math.round((config.port.canvas.completion.services + config.port.discourse.completion.services) / 2)}% |
| UI | ${config.port.canvas.completion.ui}% | ${config.port.discourse.completion.ui}% | ${Math.round((config.port.canvas.completion.ui + config.port.discourse.completion.ui) / 2)}% |
| Tests | ${config.port.canvas.completion.tests}% | ${config.port.discourse.completion.tests}% | ${Math.round((config.port.canvas.completion.tests + config.port.discourse.completion.tests) / 2)}% |
| **Total** | **${calculateOverall(config.port.canvas.completion)}%** | **${calculateOverall(config.port.discourse.completion)}%** | **${Math.round((calculateOverall(config.port.canvas.completion) + calculateOverall(config.port.discourse.completion)) / 2)}%** |

## Key Accomplishments

1. **Unified Authentication**: Implemented JWT authentication system that bridges Canvas OAuth and Discourse SSO
2. **Core Models Integration**: Successfully ported and integrated the core models from both systems
3. **API Namespacing**: Implemented properly namespaced API endpoints to avoid conflicts

## Issue Summary

${generateIssuesSummary(conflicts)}

## Integration Highlights

### Successful Integration Points

- ✅ User account synchronization
- ✅ Course-category mapping
- ✅ Discussion-topic integration
- ✅ Assignment submission workflow

### In-Progress Integration Points

- 🔄 Notification system unification
- 🔄 File management and storage
- 🔄 Analytics and reporting

## Next Steps

1. Complete the JWT authentication implementation
2. Address the ${conflicts.length} identified conflicts between source and target code
3. Implement the notification unification system
4. Improve test coverage, particularly in integration areas

## Technical Debt

The project currently carries some technical debt:

1. **Model Duplication**: Several models still have duplicate definitions
2. **Inconsistent Naming**: Variable naming conventions need standardization
3. **API Path Conflicts**: Some endpoints have potential path conflicts
4. **Missing Test Coverage**: Current test coverage is at ~${Math.round((config.port.canvas.completion.tests + config.port.discourse.completion.tests) / 2)}%

See the [full report in the Central Reference Hub](docs/central_reference_hub.md) for more details.
`;

  // Write the report
  const reportPath = path.join(__dirname, 'SUMMARY_REPORT.md');
  fs.writeFileSync(reportPath, content);
  
  console.log(`Summary report generated at ${reportPath}`);
  return reportPath;
}

/**
 * Calculate the overall completion percentage
 */
function calculateOverall(completion) {
  const sum = Object.values(completion).reduce((total, val) => total + val, 0);
  return Math.round(sum / Object.values(completion).length);
}

/**
 * Extract conflicts from markdown file
 */
function extractConflictsFromMarkdown(filePath) {
  // This is a simplified implementation
  // In a real scenario, you would parse the markdown properly
  
  const content = fs.readFileSync(filePath, 'utf8');
  const conflicts = [];
  
  // Very basic parsing
  const typeRegex = /## (.+?) Conflicts \((\d+)\)/g;
  let typeMatch;
  
  while ((typeMatch = typeRegex.exec(content)) !== null) {
    const type = typeMatch[1];
    const count = parseInt(typeMatch[2]);
    
    for (let i = 0; i < count; i++) {
      conflicts.push({ type });
    }
  }
  
  return conflicts;
}

/**
 * Generate issues summary based on conflicts
 */
function generateIssuesSummary(conflicts) {
  if (conflicts.length === 0) {
    return "No significant issues detected.";
  }
  
  // Group conflicts by type
  const byType = {};
  conflicts.forEach(conflict => {
    if (!byType[conflict.type]) byType[conflict.type] = 0;
    byType[conflict.type]++;
  });
  
  let summary = `Found ${conflicts.length} issues across ${Object.keys(byType).length} categories:\n\n`;
  
  for (const [type, count] of Object.entries(byType)) {
    summary += `- **${type}**: ${count} issues\n`;
  }
  
  return summary;
}

// Export the function
module.exports = { generateSummaryReport };

// If run directly, execute the generator
if (require.main === module) {
  generateSummaryReport()
    .then(reportPath => console.log(`Report generated at: ${reportPath}`))
    .catch(error => console.error('Failed to generate summary report:', error));
}

File: tools/unified-analyzer/src/generators/summary_report_generator.rs
use std::fs;
use std::path::Path;
use chrono::Local;

use crate::analyzers::unified_analyzer::AnalysisResult;

/// Generate summary report
pub fn generate_summary_report(result: &AnalysisResult, base_dir: &Path) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("Generating summary report...");

    // Ensure docs directory exists
    let docs_dir = base_dir.join("docs");
    if !docs_dir.exists() {
        fs::create_dir_all(&docs_dir)?;
    }

    // Create the summary report path
    let report_path = docs_dir.join("SUMMARY_REPORT.md");

    // Generate the content
    let mut content = String::new();

    // Header
    content.push_str("# LMS Project: Summary Report\n");
    content.push_str(&format!("_Generated on {}_\n\n", Local::now().format("%Y-%m-%d")));

    // Project Status
    content.push_str("## Project Status\n\n");
    content.push_str(&format!("- **Phase**: {}\n", result.project_status.phase));
    content.push_str(&format!("- **Completion**: {:.1}%\n", result.project_status.completion_percentage));
    content.push_str(&format!("- **Last Active Area**: {}\n", result.project_status.last_active_area));

    if let Some(date) = &result.project_status.estimated_completion_date {
        content.push_str(&format!("- **Estimated Completion Date**: {}\n", date.format("%Y-%m-%d")));
    }

    content.push_str("\n");

    // Implementation Metrics
    content.push_str("## Implementation Metrics\n\n");
    content.push_str("| Component | Implemented | Total | Percentage |\n");
    content.push_str("|-----------|-------------|-------|------------|\n");
    content.push_str(&format!("| Models | {} | {} | {:.1}% |\n",
        result.models.implemented,
        result.models.total,
        result.models.implementation_percentage));
    content.push_str(&format!("| API Endpoints | {} | {} | {:.1}% |\n",
        result.api_endpoints.implemented,
        result.api_endpoints.total,
        result.api_endpoints.implementation_percentage));
    content.push_str(&format!("| UI Components | {} | {} | {:.1}% |\n",
        result.ui_components.implemented,
        result.ui_components.total,
        result.ui_components.implementation_percentage));
    content.push_str(&format!("| Integration Points | {} | {} | {:.1}% |\n",
        result.integration.implemented_points,
        result.integration.total_points,
        result.integration.implementation_percentage));

    content.push_str("\n");

    // Code Quality Metrics
    content.push_str("## Code Quality Metrics\n\n");

    for (metric, value) in &result.code_quality.metrics {
        content.push_str(&format!("- **{}**: {:.1}\n", metric, value));
    }

    content.push_str(&format!("- **Test Coverage**: {:.1}%\n", result.tests.coverage));

    content.push_str("\n");

    // Testing Metrics
    content.push_str("## Testing Metrics\n\n");
    content.push_str(&format!("- **Total Tests**: {}\n", result.tests.total));
    content.push_str(&format!("- **Passing Tests**: {}\n", result.tests.passing));

    // Calculate pass rate
    let pass_rate = if result.tests.total > 0 {
        (result.tests.passing as f32 / result.tests.total as f32) * 100.0
    } else {
        0.0
    };

    content.push_str(&format!("- **Pass Rate**: {:.1}%\n", pass_rate));

    content.push_str("\n");

    // Architecture
    content.push_str("## Architecture\n\n");
    content.push_str("### Frameworks\n\n");

    for framework in &result.architecture.frameworks {
        content.push_str(&format!("- {}\n", framework));
    }

    content.push_str("\n");

    content.push_str("### Design Patterns\n\n");

    for pattern in &result.architecture.design_patterns {
        content.push_str(&format!("- {}\n", pattern));
    }

    content.push_str("\n");

    // Blockchain
    content.push_str("## Blockchain\n\n");
    content.push_str(&format!("- **Status**: {}\n", result.blockchain.implementation_status));
    content.push_str("- **Features**:\n");

    for feature in &result.blockchain.features {
        content.push_str(&format!("  - {}\n", feature));
    }

    content.push_str("\n");

    // Recommendations
    content.push_str("## Recommendations\n\n");

    for recommendation in &result.recommendations {
        content.push_str(&format!("- **{}**: {}\n", recommendation.area, recommendation.description));
    }

    content.push_str("\n");

    // Next Steps
    content.push_str("## Next Steps\n\n");
    content.push_str("1. Address high priority technical debt\n");
    content.push_str("2. Implement remaining API endpoints\n");
    content.push_str("3. Increase test coverage\n");
    content.push_str("4. Complete model implementations\n");
    content.push_str("5. Improve documentation\n");

    // Write to file
    fs::write(&report_path, content)?;

    println!("Summary report generated at: {:?}", report_path);

    Ok(())
}
